### PR TITLE
Pull Image IDs from SSM Parameter Store

### PR DESCRIPTION
**Pull request by PettitWesley**
For the feature launched today: [ECS AMI Metadata in SSM](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/retrieve-ecs-optimized_AMI.html)

### CODE REVIEW

**Comment by PettitWesley** [comment]
Testing Info:

##### Tested that Fargate functionality still works, since I touched that part of the code:
```
$ ecs-cli up --cluster-config fargate-test                       
INFO[0000] Validated params                             
INFO[0000] Created cluster                               cluster=FURRGATE-RETURNS region=us-east-1
INFO[0001] Waiting for your cluster resources to be created... 
INFO[0001] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-092b76309c8dc3f23
Subnet created: subnet-0c44a606fcbba558f
Subnet created: subnet-0ae5234d45d85729b
Cluster creation succeeded.
```

### Tested new stuff in a bunch of regions

##### sa-east-1
```
$ ecs-cli up --capability-iam --size 2 --instance-type t2.medium --region sa-east-1 --launch-type EC2 --force --cluster Metadata
WARN[0000] You will not be able to SSH into your EC2 instances without a key pair. 
INFO[0001] Using recommended Amazon Linux AMI with ECS Agent 1.17.3 and Docker version 17.12.1-ce 
INFO[0003] Created cluster                               cluster=Metadata region=sa-east-1
INFO[0004] Waiting for your cluster resources to be created... 
INFO[0005] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0067] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0129] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-0293f60ab85ebf3d2
Security Group created: sg-0b0a0b3631a29d880
Subnet created: subnet-03a458f4b92c9de31
Subnet created: subnet-02b94721350cd6493
Cluster creation succeeded.
```

##### ap-northeast-1

```
$ ecs-cli up --capability-iam --size 2 --instance-type t2.medium --region ap-northeast-1 --launch-type EC2 --force --cluster Metadata
WARN[0000] You will not be able to SSH into your EC2 instances without a key pair. 
INFO[0001] Using recommended Amazon Linux AMI with ECS Agent 1.17.3 and Docker version 17.12.1-ce 
INFO[0001] Created cluster                               cluster=Metadata region=ap-northeast-1
INFO[0002] Waiting for your cluster resources to be created... 
INFO[0002] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0063] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0125] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-0854f0c155159efa1
Security Group created: sg-0e66d26b78b39c0ab
Subnet created: subnet-069ae508ffff7d87f
Subnet created: subnet-0de4f04aa143f4f96
Cluster creation succeeded.
```

##### us-east-2

```
$ ecs-cli up --capability-iam --size 2 --instance-type t2.medium --region us-east-2 --launch-type EC2 --force --cluster Metadata
WARN[0000] You will not be able to SSH into your EC2 instances without a key pair. 
INFO[0000] Using recommended Amazon Linux AMI with ECS Agent 1.17.3 and Docker version 17.12.1-ce 
INFO[0001] Created cluster                               cluster=Metadata region=us-east-2
INFO[0001] Waiting for your cluster resources to be created... 
INFO[0001] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0062] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0123] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-0aca60590a133160e
Security Group created: sg-06f0c5e55dffeb067
Subnet created: subnet-018319ac4e494ce73
Subnet created: subnet-090c928ff07c7f238
Cluster creation succeeded.
```

##### eu-central-1

```
$ ecs-cli up --capability-iam --size 2 --instance-type t2.medium --region eu-central-1 --launch-type EC2 --force --cluster Metadata
WARN[0000] You will not be able to SSH into your EC2 instances without a key pair. 
INFO[0001] Using recommended Amazon Linux AMI with ECS Agent 1.17.3 and Docker version 17.12.1-ce 
INFO[0002] Created cluster                               cluster=Metadata region=eu-central-1
INFO[0003] Waiting for your cluster resources to be created... 
INFO[0003] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0065] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0127] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-0b2dbf7ec407e4241
Security Group created: sg-07c2ca89cabfc15e5
Subnet created: subnet-0d511738a951065ca
Subnet created: subnet-060eaa6616acc16d7
Cluster creation succeeded.
```

##### ap-southeast-2

```
$ ecs-cli up --capability-iam --size 2 --instance-type t2.medium --region ap-southeast-2 --launch-type EC2 --force --cluster Metadata
WARN[0000] You will not be able to SSH into your EC2 instances without a key pair. 
INFO[0001] Using recommended Amazon Linux AMI with ECS Agent 1.17.3 and Docker version 17.12.1-ce 
INFO[0002] Created cluster                               cluster=Metadata region=ap-southeast-2
INFO[0003] Waiting for your cluster resources to be created... 
INFO[0004] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0066] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0128] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-05c6b460e8c83bf17
Security Group created: sg-06fdcff1b2dab6bf2
Subnet created: subnet-07d6b1d9d9e7ef6e0
Subnet created: subnet-0a586b8a96e45e391
Cluster creation succeeded.

```

File: ecs-cli/modules/clients/aws/cloudformation/params_test.go [hunk 1]
```diff
@@ -59,8 +59,8 @@ func TestAddAndValidate(t *testing.T) {
 	}
 
 	paramsMap := cfnParams.Get()
-	if len(requiredParameterNames) != len(paramsMap) {
-		t.Errorf("Mismatch in number of keys in params map. %d != %d", len(requiredParameterNames), len(paramsMap))
+	if len(paramsMap) != 2 {
```
Comment (SoManyHs) [inline ecs-cli/modules/clients/aws/cloudformation/params_test.go]: What is the significance of 2? Is there a more descriptive way we can make this assertion?

File: ecs-cli/modules/cli/cluster/cluster_app.go [hunk 2]
```diff
@@ -159,7 +159,7 @@ func validateCommaSeparatedParam(cfnParams *cloudformation.CfnStackParams, param
 	return false
 }
 
-func createCluster(context *cli.Context, ecsClient ecsclient.ECSClient, cfnClient cloudformation.CloudformationClient, cliParams *config.CLIParams) error {
+func createCluster(context *cli.Context, ecsClient ecsclient.ECSClient, cfnClient cloudformation.CloudformationClient, ssmClient ssm.Client, cliParams *config.CLIParams) error {
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/cluster/cluster_app.go]: No op: This is fine, but note for future self to maybe consider converting some of the arguments here into a parameter object so we don't have quite so many in the method signature.

File: ecs-cli/modules/clients/aws/cloudformation/params_test.go (hunk 1 above)
Comment (PettitWesley) [reply ecs-cli/modules/clients/aws/cloudformation/params_test.go]: 2 is just the number of parameters that were added in the test. So its checking if the 2 calls to the parameter add method actually added 2 parameters.

Comment (PettitWesley) [reply ecs-cli/modules/clients/aws/cloudformation/params_test.go]: I can make 2 a const with a descriptive name tho

**Comment by PettitWesley** [comment]
@allisaurus Did more testing:

#### cn-north-1
```
$ ecs-cli up --capability-iam --size 2 --instance-type t2.medium --region cn-north-1 --launch-type EC2 --force --cluster Metadata
WARN[0000] You will not be able to SSH into your EC2 instances without a key pair. 
INFO[0001] Using recommended Amazon Linux AMI with ECS Agent 1.17.3 and Docker version 17.12.1-ce 
INFO[0002] Created cluster                               cluster=Metadata region=cn-north-1
INFO[0003] Waiting for your cluster resources to be created... 
INFO[0003] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0065] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0127] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-ae3044ca
Security Group created: sg-d0c026b7
Subnet created: subnet-1fe45a7b
Subnet created: subnet-8e4971f9
Cluster creation succeeded.
```

#### ca-central-1 with a KeyPair

```
$ ecs-cli up --capability-iam --size 2 --instance-type t2.medium --region ca-central-1 --keypair MyFirstKeyPair  --launch-type EC2 --force --cluster Metadata 
INFO[0000] Using recommended Amazon Linux AMI with ECS Agent 1.17.3 and Docker version 17.12.1-ce 
INFO[0001] Created cluster                               cluster=Metadata region=ca-central-1
INFO[0001] Waiting for your cluster resources to be created... 
INFO[0002] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0063] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
INFO[0124] Cloudformation stack status                   stackStatus=CREATE_IN_PROGRESS
VPC created: vpc-0faa8f3fa55790b90
Security Group created: sg-0b8c73cc0c477927d
Subnet created: subnet-00475c6961815d20a
Subnet created: subnet-0578a5449faf69cff
Cluster creation succeeded.
```

File: ecs-cli/Gopkg.lock [hunk 3]
```diff
@@ -193,6 +193,6 @@
 [solve-meta]
   analyzer-name = "dep"
   analyzer-version = 1
-  inputs-digest = "b911e31ad580796071e4d98e7b5d2d0eafb01d15c968aba682e83ca05659e94d"
+  inputs-digest = "ef0a39362eca40e396dfd56b7fec991eeac9d4f2632c40985642d9b34fad7b9b"
```
[the remark this answers is not in the document]
Comment (PettitWesley) [reply ecs-cli/Gopkg.lock]: fixed

**Comment by SoManyHs** [comment]
LGTM 👍

**Comment by kaihendry** [comment]
Is there some reason why it's better to pull out the IDs from the parameter store instead of a well defined public https endpoint?

Perhaps the parameter store ultimately is a public endpoint, I'm not sure?

**Comment by PettitWesley** [comment]
@kaihendry The AMI IDs are pulled using the [ssm:GetParameter](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_GetParameter.html) API, and the AWS SDK Go will make this call to the https endpoint for the SSM service. The [ECS Optimized AMI Metadata](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/retrieve-ecs-optimized_AMI.html) in SSM is controlled by ECS, and you can be confident that we will keep it secure and up to date. 

If you have any more concerns with the ECS CLI using this new feature, please let me know.

**Pull request merged**